core/arch/arm/plat-rzn1/main.c

```diff
@@ -72,3 +85,36 @@ static TEE_Result rzn1_tz_init(void)
 }
 
 service_init(rzn1_tz_init);
+
+#ifdef CFG_BOOT_CM3
+static TEE_Result rzn1_cm3_start(void)
+{
+	vaddr_t cm3_pwrctrl_reg = 0;
+	vaddr_t cm3_pwrstat_reg = 0;
+	uint64_t timeout_ack = timeout_init_us(IDLE_ACK_TIMEOUT_US);
+
+	cm3_pwrctrl_reg = core_mmu_get_va(SYSCTRL_PWRCTRL_CM3, MEM_AREA_IO_SEC,
+					  sizeof(uint32_t));
+	cm3_pwrstat_reg = core_mmu_get_va(SYSCTRL_PWRSTAT_CM3, MEM_AREA_IO_SEC,
+					  sizeof(uint32_t));
+
+	/* Master Idle Request to the interconnect for CM3 */
+	io_clrbits32(cm3_pwrctrl_reg, SYSCTRL_PWRCTRL_CM3_MIREQ_A);
+
+	/* Wait for Master Idle Request Acknowledge for CM3 */
+	while (!timeout_elapsed(timeout_ack))
+		if (!(io_read32(cm3_pwrstat_reg) & SYSCTRL_PWRSTAT_CM3_MIRACK_A))
+			break;
+
+	if (io_read32(cm3_pwrstat_reg) & SYSCTRL_PWRSTAT_CM3_MIRACK_A)
+		return TEE_ERROR_GENERIC;
```

Comment:
> @jforissier, I wonder if `call_initcalls()` should panic of init function failure.

No, the current scheme is more flexible and there is not much to gain from changing it I think:
- If we were to replace `panic("Some error text")` calls with `return TEE_ERROR_xxx;` we would need to add a `EMSG()` to keep the message
- Panicking inside the failed function is a bit better for the stack dump (if `panic()` is after `call->func()` you need to identify the function from its address in the error message)
- Some init functions currently may fail without panicking (for example `platform_init()` in `core/arch/arm/plat-rockchip/platform.c`), it would look slightly weird to hve them return success after a non-fatal error

Reply:
ok. makes sense. let's keep as-is.

Reply:
I changed it to call `panic()`